feat(STN-174):  stylelint codeclimate update

silverli opened the pull request:
# READY FOR REVIEW

## Summary
- adds stylelint to codeclimate.yml and removes scss-lint
- adds stylelintrc rules based on old plugins we were pulling in
- fixes all stylelint rules in other themes
- adds .stylelintignore so that codeclimate doesn't analyze entire codebase with stylelint...
- takes out scss-lint comments from our scss code
- gets codeclimate passing with new set up with develop branch ✅ https://codeclimate.com/github/SU-HSDO/suhumsci/pull/555 
- adds backstop js work here with new tasks

✅ run backstop js on old theme (looks good)
✅  add compiled CSS for other themes - there actually are no changes, so that's a great sign!

## Steps to Test
1. run `npm run test` in humsci_basic, ensure tests pass
2. run `npm i` in the root folder (you might need to delete the lock file or node_modules
3. run `npm run test` in the main root folder and make sure it passes
4. try adding a failing scss style in another theme scss folder, see that the stylelint fails
5. check out the humsci_basic readme and try out backtop js. you will need to run `npm i`

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

Commit:
feat(STN-174): first work at switching from scss-lint to styelint

- adds new packages to package.json
- adds linting rules to root folder
- can now run "npm run lint:sass" from root for all scss in project
- fixes a bunch of stylelint issues in other sites themes
- TODO: still need to add gruntfil watch task or something with the new
  linting
- TODO: check if codeclimate works against develop
- TODO: get rid of scss-lint specific comments in code
- TODO: fix remaining stylelint issues

Commit:
feat(STN-174): manually adds all required stylelint rules

Commit:
feat(STN-174): get rid of all sparkbox linting dependencies"

Commit:
feat(STN-174): delete humsci_basic stylelint

Commit:
feat(STN-174): fixes stylelint issues that are currently present in
other themes

- also adds new tasks to main package json
- gets rid of things ignoring humsci_basic scss files in the main
  gruntfile

Commit:
refactor(STN-174): gets rid of scss-lint disable comments

Commit:
feat(STN-174): adds root .styelintignore

Commit:
feat(STN-174): adds root stylelint to humsci basic config

kaseybon committed:
feat: add backstopjs for visual regression testing

Commit:
feat(STN-174): small typo

Commit:
refactor(STN-174): backstop js for humsci basic uses and readme

heatherdesigns commented on docroot/themes/humsci/humsci_basic/readme.md:
````diff
@@ -45,6 +45,27 @@ We use [stylelint](https://stylelint.io/) to lint all of our Sass code to mainta
+  }
+]
+```
+1. Run `npm backstop:init` to generate reference images, in our case reference is production:
````
`npm backstop:init` did not work for me. Although `npm run backstop:init` did. Can we update this?

silverli replied:
oops!! yes I'll update

heatherdesigns commented on .stylelintrc:
```diff
@@ -0,0 +1,139 @@
+  "plugins": [
+    "stylelint-scss" # plugin for SCSS specific rules
+  ],
+  "rules": {
```
Looking at the rules my first instinct is to ask if we can alphabetize them? But I remember you said something about the importance of the order in which the rules are declared. Can we add a comment explaining that?

silverli replied:
yep I can alphabetize, no problem

heatherdesigns commented on docroot/themes/humsci/humsci_basic/backstop/scenarios.json:
```diff
@@ -0,0 +1,102 @@
+      "requireSameDimensions": true
+    },
+    {
+      "label": "QA Content",
```
The scenarios for Econ did not appear in the results of my BackstopJS Report. I only saw the files for the default site. According to the `readme` instructions, new scenarios are added to the `backstop/scenarios.json` file. 
 I'm curious as to what this json file used for?

silverli replied:
Yeah.. I was hoping we'd be able to import a separate json file into here, but it wasn't possible, and now a separate file is just confusing. I'll delete scenarios.json and add all scenarios here.

heatherdesigns commented on docroot/themes/humsci/su_humsci_theme/patterns/molecules/accordion/scss/accordion.scss:
```diff
@@ -9,8 +9,8 @@ details {
-      &:after {
-        content: "\f068"
+      &::after {
+        content: "\f068";
```
Oh wow! I'm surprised the other linter didn't pick up on the missing semicolon. That seems kinda like it would be a big deal. 🤷

silverli replied:
right?!! 😂

heatherdesigns reviewed (changes requested):
Excellent work parsing through all of files and making the necessary updates!! I left a request for a couple minor changes and and I have a couple questions for clarification.

![highest of fives](https://user-images.githubusercontent.com/12678977/79137319-7fb3d800-7d80-11ea-86e7-ba3cd40e20c7.gif)

silverli commented on Gruntfile.js:
```diff
@@ -5,7 +5,6 @@ module.exports = function (grunt) {
 
   function getIncludeFiles() {
     var patterns = [
-      'docroot/themes/humsci/!(humsci_basic)/**/*.scss',
```
@heatherdesigns I realized I need to keep these rules in here to exclude humsci basic. the root linting task will still look at everything, but the build task cannot because we have special stuff going on in humsci basic

bryanbraun replied:
I agree with this. It would be cool if, at some point, we could get everything on the same build task. But that's something we can think about another day.

Commit:
feat(STN-174): revise readme and backstop.json

Commit:
feat(STN-174): alphabetize .stylelintrc

Commit:
feat(STN-174): update readme and fix Gruntfile

- we actually cannot have the main gruntfile build humsci_basic because
  that build is so unique
- so I added those ignores for humsci_basic back
- I also had inadvertently broken the build by removing those
  statements - yikes
- so, now, you are able to lint the whole project from the root, but you
  cannot build humsci_basic from the root (but everything else, yes)

Commit:
feat(STN-174): add compiled CSS for small theme changes

- the css is only different because of changes like making the use of
  colons and psuedo elements consistent, like adding `&::after` instead
of `&:after` or adding a leading zero to a number

bryanbraun commented on package.json:
```diff
@@ -10,7 +10,9 @@
     "build:sass": "grunt compile",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "build": "npm run build:sass",
+    "lint:sass": "stylelint 'docroot/themes/humsci/**/{scss,src,color}/**/*.{scss,sass}' --config .stylelintrc",
```
Hey @pookmish, this PR adds a task for linting the whole codebase, to the root package.json of this project ([like we discussed before](https://github.com/SU-HSDO/suhumsci/pull/487#issuecomment-586488932)).

Once packages are installed (`npm install`) you can run the linting with `npm lint:sass` (or `npm test`). Is this the format you'd want this in? Alternatively, we could do this with a grunt task. Let me know what you think is best. 👍

bryanbraun replied:
Mike confirmed with me that either works for him. 👍

bryanbraun commented on docroot/themes/humsci/humsci_basic/readme.md:
```diff
@@ -25,15 +25,14 @@ This theme contains its own node module dependencies and build system which is s
 ## Builds
 
-Frontend assets are built using the Grunt task runner, but are run using npm scripts as shortcuts. Assets are compiled to a `dist/` directory with a stylesheet generated for each sub-theme.
+Frontend assets are built using the Grunt task runner, but are run using npm scripts as shortcuts. CSS assets are compiled to their respective child theme `css/` directory. JS assets are compiled to the `src/js/scripts.js` file.
```
Great docs update. 👍

bryanbraun reviewed (approved):
These changes look great, and I didn't see any regressions in the visual diffing, so I feel good about this!
